### ISSUE TITLE
Copula model delivers poor performance on categorical data

### ISSUE THREAD

**Issue by mkffl** [sdv-dev/SDV#310]
## Problem description

I can't seem to get decent results when modelling fields with more than one category. Apologies if I am missing an obvious trick here -- I read the docs and scouted existing issues but couldn't find references to my issue.

To clarify, I think of category as a group of similar individuals that share common properties. In some models, e.g. bayesian hierarchical models, a group can be assumed to come from the same distribution. The sdv docs suggest that categories can be modelled using the `categorical` Type in the model definition.

## Expected behaviour

To set some minimum expectations to work towards:
- Have at least one sample in each category
- An interquartile range that's not too dissimilar -- below you will see that category 2's synthetic 75th percentile is c.4x the original threshold
- Some stability when original distributions are skewed (below I set reasonably small skewness params but when cranked up from 2 to 10, the model really struggles)

Note that **I am able to achieve really good results** if I fit a separate model on each category. This suggests that I probably don't configure the model correctly (again, sorry).

## Example

To illustrate the low quality of my copula model output, I provide a minimalist example below. Results are really off-the-mark, with some categories simply not being sampled, and others having vastly different statistics than the original. 

```python
import numpy as np
import pandas as pd
from scipy import stats
import seaborn as sns
from typing import Tuple
from toolz import pipe
from toolz.curried import *
from sdv.tabular import GaussianCopula

# Generate some "real" data to fit a Gaussian Copula model on

# Set some mild skewness values to make the data look real
skewness_param1 = 2

skewness_param2 = -1

skewness_param3 = 2

skewness_param4 = -2

cateory_sample_size1 = 100

cateory_sample_size2 = 80

cateory_sample_size3 = 30

cateory_sample_size4 = 150

def sample_data(params: Tuple[int]):
    return stats.skewnorm(params[0]).rvs(params[1])*10


# Create the data
data_params = [
    (skewness_param1, cateory_sample_size1),
    (skewness_param2, cateory_sample_size2),
    (skewness_param3, cateory_sample_size3),
    (skewness_param4, cateory_sample_size4)
]

data = pipe(data_params,
     map(sample_data),
     reduce(lambda l, r: np.concatenate([l,r])),
     pd.DataFrame
    )

# Add labels and primary key

labels = ["cat1"]*cateory_sample_size1 + ["cat2"]*cateory_sample_size2 + ["cat3"]*cateory_sample_size3 + ["cat4"]*cateory_sample_size4

data["label"] = labels

data["id"] = list(range(len(labels)))

data.columns = ["value", "label", "id"]

# Fit a model and get some samples

model = GaussianCopula(
        primary_key='id',
        field_types={
            "value": {"type":"numerical", 
                       "subtype": "float"
                      },
            "label": {"type":"categorical"
                      },
            "id": {"type":"id", "subtype":"integer"
                       }
        }
    )

model.fit(data)

model_samples = model.sample(len(data))
```

### 1. Value counts

`model_samples.label.value_counts()`

```
cat4    152
cat1    114
cat2     94
Name: label, dtype: int64
```
`category 3` is missing. OK, we may need more samples:

```python
model_more_samples = model.sample(5*len(data))

model_more_samples.label.value_counts()
```

```
cat4    764
cat1    592
cat2    444
Name: label, dtype: int64
```

Still missing... Let's eyeball the results with a boxplot

### 2. Distribution plots / tables
```python
order_label = data.label.value_counts().sort_index().index

sns.boxplot(x="label", y="value", data=data, order=order_label)
```

![image](https://user-images.githubusercontent.com/10297799/106797968-3b6a2f00-6655-11eb-9153-89f492eb9674.png)

```python
sns.boxplot(x="label", y="value", data=model_more_samples, order=order_label)
```

![image](https://user-images.githubusercontent.com/10297799/106797997-4755f100-6655-11eb-8fcb-f7c87bced8cd.png)

Not amazing, especially for category 2. Its 25-75th percentile range is not in line with the original data:

![image](https://user-images.githubusercontent.com/10297799/106798470-ebd83300-6655-11eb-8ad9-c7f392c622cf.png)

**Comment by csala** [comment]
Thanks for this detailed report @mkffl !

It looks indeed like a configuration issue.

For example, if I run your example using `categorical_transformer='categorical_fuzzy'` I get a distribution of values much closer to the original one:

```python3
model = GaussianCopula(
    primary_key='id',
    categorical_transformer='categorical_fuzzy',
)
model.fit(data)
model.sample().groupby('label').describe()['value']
```

![image](https://user-images.githubusercontent.com/6129283/106894804-b7b15080-66ef-11eb-9f1d-de7b2c291a6c.png)

But in this case the distribution of the `value` conditioned on the `label` is not that close to the original data distributions either:

```python3
data.groupby('label').describe()['value']
```

![image](https://user-images.githubusercontent.com/6129283/106894969-f21aed80-66ef-11eb-9748-d4a7df54669d.png)


On the other side, if instead I change `default_distribution` to `univariate`, which would also get the `gaussian_kde` distribution as a possible candidate, I get again a good distribution of `label` categories, and a distribution of `value` conditioned on those much closer to what it should be:

```python3
model = GaussianCopula(
    primary_key='id',
    categorical_transformer='categorical_fuzzy',
)
model.fit(data)
model.sample().groupby('label').describe()['value']
```

![image](https://user-images.githubusercontent.com/6129283/106894832-c0a22200-66ef-11eb-8996-ad72a8b668cc.png)

**Comment by mkffl** [comment]
Thanks for your explanations, which I very much appreciate.  Using `categorical_transformer` in the model helps with sampling.

Your code snippets do not seem to match your text. Do you set `default_distribution` to `univariate`?

I think I am almost there but there's still something off... I am certain I would still get better results by fitting a standalone model on each category.

```python

model = GaussianCopula(
        primary_key='id',
        categorical_transformer='categorical',
        default_distribution='univariate',
        field_types={
            "value": {"type":"numerical", 
                       "subtype": "float"
                      },
            "label": {"type":"categorical"
                      },
            "id": {"type":"id", "subtype":"integer"
                       }
        }
    )

model.fit(data)

model_samples = model.sample(2*len(data))

model_samples.label.value_counts()
```

```
cat4    325
cat1    194
cat2    162
cat3     39
Name: label, dtype: int64
```

Category 2's mean looks very far away from the truth. 

<img width="580" alt="sdv" src="https://user-images.githubusercontent.com/10297799/106951192-5a38f600-6727-11eb-8438-a28a035841bc.png">




Note that I now apply a loc-transform to push the four categories' means apart.

```python
def sample_data(params: Tuple[int]):
    return (stats.skewnorm(params[0]).rvs(params[1])*10+
            np.random.randint(low=-3, high=3))
```

**Comment by npatki** [comment]
It's been several years since this issue was filed, so I'm marking it as stale. In the last few years, we've made significant improvements to pre-processing that should be resulting in significantly better categorical data quality.

This includes:
- Applying the [LabelEncoder](https://docs.sdv.dev/rdt/transformers-glossary/categorical/labelencoder)(with noise) to preprocess each categorical and boolean column. In our experimentation, this actually outperforms categorical fuzzy (since renamed to Frequency Encoder).
- Allowing you more control over [customizing the preprocessing](https://docs.sdv.dev/sdv/single-table-data/modeling/synthetic-data-workflow/transform-and-anonymize) so that you can add the methods that make sense for you. 

If you're continuing to experience data quality problems with categorical data, please feel free to file a new issue!
